src/lir.rs

```diff
@@ -0,0 +1,452 @@
+    }
+}
+
+#[derive(Debug)]
+pub enum LirInstruction {
+    LogicalOperator(LirLogicalOperator),
+    Predicate(Predicate),
+}
+
+impl LirInstruction {
+    #[inline]
+    pub fn as_predicate(&self) -> &Predicate {
+        match &self {
+            LirInstruction::LogicalOperator(_ops) => {
+                panic!("Call as_predicate on LogicalOperator Operand, LirProgram is wrong.")
+            }
+            LirInstruction::Predicate(p) => p,
+        }
+    }
+}
+
+#[inline]
+pub(crate) fn is_operator(instruction: &LirInstruction) -> bool {
+    match instruction {
+        LirInstruction::LogicalOperator(_op) => true,
+        LirInstruction::Predicate(_p) => false,
+    }
+}
+
+#[derive(Debug)]
+pub enum LirLogicalOperator {
+    And,
+    Or,
+    Not,
+}
+
+pub trait Translate {
+    type Output;
+    fn translate(&self) -> Self::Output;
+}
+
+#[cfg(debug_assertions)]
+trait CountSize {
+    type Output;
+    fn count_size(&self) -> Self::Output;
+}
+#[cfg(debug_assertions)]
+#[derive(Debug, Default)]
+struct ExpressionInsBytes {
+    ins_number: usize,
+    ins_bytes: usize,
+}
+#[cfg(debug_assertions)]
+impl ExpressionInsBytes {
+    pub fn new() -> Self {
+        Self {
+            ins_number: 0,
+            ins_bytes: 0,
+        }
+    }
+}
+#[cfg(debug_assertions)]
+fn expression_count_heler(exp: &Expression, counter: &mut ExpressionInsBytes) {
+    use crate::ast::{Expression, LogicalExpression};
+    use std::mem;
+    counter.ins_bytes += mem::size_of::<Expression>();
+    match exp {
+        Expression::Logical(l) => {
+            counter.ins_number += 1;
+            counter.ins_bytes += mem::size_of::<LogicalExpression>();
+            match l.as_ref() {
+                LogicalExpression::And(l, r) => {
+                    expression_count_heler(l, counter);
+                    expression_count_heler(r, counter);
+                }
+                LogicalExpression::Or(l, r) => {
+                    expression_count_heler(l, counter);
+                    expression_count_heler(r, counter);
+                }
+                LogicalExpression::Not(r) => {
+                    expression_count_heler(r, counter);
+                }
+            }
+        }
+        Expression::Predicate(_p) => {}
+    }
+}
+#[cfg(debug_assertions)]
+impl CountSize for Expression {
+    type Output = ExpressionInsBytes;
+    fn count_size(&self) -> Self::Output {
+        let mut counter = ExpressionInsBytes::new();
+        expression_count_heler(self, &mut counter);
+        counter
+    }
+}
+
+impl Translate for Expression {
+    type Output = LirProgram;
+    fn translate(&self) -> Self::Output {
+        let mut lir = LirProgram::new();
+        lir_translate_helper(self, &mut lir);
+        lir.instructions.shrink_to_fit(); // shrink the memory
+        #[cfg(debug_assertions)]
+        {
+            let ast_counter = self.count_size();
+            ast_counter.ins_bytes;
+            ast_counter.ins_number;
+        }
```

Comment:
Please comment out the code without any effect. Clippy is unhappy with that. Otherwise, use `let _ = ...` to bypass the check (But it makes no sense to leave unused statements here.)

Reply:
fixed